File: javascript/triangle/triangle.js
/**
 * Returns a distinct array of elements
 * @param arr
 * @returns {Array}
 */
function distinctArray(arr) {
    return Object.keys(arr.reduce(function (map, len) {
        map[len] = true;
        return map;
    }, {}));
}

/**
 * @param sides
 * @throws Error if the sides violate the triangle inequality principle
 */
function validateSides(sides) {
    sides.sort();
    if (sides.length != 3 || sides[0] < 0 || (distinctArray(sides).length === 1 && sides[0] === 0)
        || ((sides[0] + sides[1] ) <= sides[2])) {
        throw new Error("Invalid lengths provided");
    }
}

// name -> # of distinct lengths
var TYPES = {
    equilateral: 1,
    isosceles: 2,
    scalene: 3
};

module.exports = function () {
    var sides = Array.prototype.slice.call(arguments, 0, 3);
    return {
        kind: function () {
            validateSides(sides);
            var uniqueSides = distinctArray(sides).length;
            return Object.keys(TYPES).reduce(function (current, type) {
                return current ? current : (TYPES[type] === uniqueSides ? type : null);
            }, null);
        }
    };
};

File: elixir/triangle/lib/triangle.ex
defmodule Triangle do
  @type kind :: :equilateral | :isosceles | :scalene

  @doc """
  Return the kind of triangle of a triangle with 'a', 'b' and 'c' as lengths.
  """
  @spec kind(number, number, number) :: {:ok, kind} | {:error, String.t()}
  def kind(a, b, c) do
    with :ok <- validate(a, b, c) do
      case [a, b, c] |> Enum.uniq() |> length() do
        1 -> {:ok, :equilateral}
        2 -> {:ok, :isosceles}
        3 -> {:ok, :scalene}
      end
    end
  end

  defp validate(a, b, c) when a > 0 and b > 0 and c > 0 do
    if a + b >= c and a + c >= b and b + c >= a do
      :ok
    else
      {:error, "side lengths violate triangle inequality"}
    end
  end

  defp validate(_, _, _), do: {:error, "all side lengths must be positive"}
end
